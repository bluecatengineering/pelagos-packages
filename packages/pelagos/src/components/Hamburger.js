import PropTypes from 'prop-types';

import './Hamburger.less';

/** A hamburger button. */
const Hamburger = ({active, onClick, ...props}) => (
	<button {...props} className="Hamburger" type="button" onClick={onClick}>
		<span className={`Hamburger__inner${active ? ' Hamburger--active' : ''}`} />
	</button>
);

Hamburger.propTypes = {
	/** Whether the button is active. */
	active: PropTypes.bool,
	/** Function invoked when the button is clicked. */
	onClick: PropTypes.func,
};

export default Hamburger;
