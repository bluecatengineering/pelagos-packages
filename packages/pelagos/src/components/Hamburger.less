@import '../../less/focus';
@import '../../less/utils';

@layer pelagos {
	.Hamburger {
		@icon-button();
		display: inline-flex;
		position: relative;
		width: 48px;
		height: 48px;
		justify-content: center;
	}

	.Hamburger__inner,
	.Hamburger__inner::before,
	.Hamburger__inner::after {
		position: absolute;
		width: 20px;
		height: 3px;
		transition: transform 0.15s ease;
		border-radius: 3px;
		background-color: currentcolor;
	}

	.Hamburger__inner::before,
	.Hamburger__inner::after {
		display: block;
		content: '';
	}

	.Hamburger__inner {
		top: 50%;
		display: block;
		margin-top: -2px;
	}

	.Hamburger__inner::before {
		top: -7px;
		transition:
			top 75ms ease 0.12s,
			opacity 75ms ease;
	}

	.Hamburger__inner::after {
		bottom: -7px;
		transition:
			bottom 75ms ease 0.12s,
			transform 75ms cubic-bezier(0.55, 0.055, 0.675, 0.19);
	}

	.Hamburger--active {
		transition-delay: 0.12s;
		transition-timing-function: cubic-bezier(0.215, 0.61, 0.355, 1);
		transform: rotate(45deg);
	}

	.Hamburger--active::before {
		top: 0;
		transition:
			top 75ms ease,
			opacity 75ms ease 0.12s;
		opacity: 0;
	}

	.Hamburger--active::after {
		bottom: 0;
		transition:
			bottom 75ms ease,
			transform 75ms cubic-bezier(0.215, 0.61, 0.355, 1) 0.12s;
		transform: rotate(-90deg);
	}
}
