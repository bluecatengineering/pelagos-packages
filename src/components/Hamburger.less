@import '../../less/colors';

.Hamburger {
	position: relative;
	width: 44px;
	height: 44px;
	color: @azure;
	align-items: center;
	cursor: pointer;
	transition: color 0.3s ease-out;
}

.Hamburger:hover {
	color: @sea;
}

.Hamburger:active {
	color: @ocean;
}

.Hamburger:focus {
	color: @sea;
	outline: 2px solid @bondi;
	outline-offset: 0;
}

.Hamburger__inner,
.Hamburger__inner::before,
.Hamburger__inner::after {
	position: absolute;
	width: 20px;
	height: 3px;
	transition: transform 0.15s ease;
	border-radius: 3px;
	background-color: currentColor;
}

.Hamburger__inner::before,
.Hamburger__inner::after {
	display: block;
	content: '';
}

.Hamburger__inner {
	top: 50%;
	display: block;
	margin-top: -2px;
}

.Hamburger__inner::before {
	top: -7px;
	transition: top 75ms ease 0.12s, opacity 75ms ease;
}

.Hamburger__inner::after {
	bottom: -7px;
	transition: bottom 75ms ease 0.12s, transform 75ms cubic-bezier(0.55, 0.055, 0.675, 0.19);
}

.Hamburger--active {
	transition-delay: 0.12s;
	transition-timing-function: cubic-bezier(0.215, 0.61, 0.355, 1);
	transform: rotate(45deg);
}

.Hamburger--active::before {
	top: 0;
	transition: top 75ms ease, opacity 75ms ease 0.12s;
	opacity: 0;
}

.Hamburger--active::after {
	bottom: 0;
	transition: bottom 75ms ease, transform 75ms cubic-bezier(0.215, 0.61, 0.355, 1) 0.12s;
	transform: rotate(-90deg);
}
