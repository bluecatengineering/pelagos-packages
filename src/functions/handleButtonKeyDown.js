export default (event) =>
	!event.shiftKey && !event.ctrlKey && !event.altKey && !event.metaKey && (event.keyCode === 13 || event.keyCode === 32)
		? (event.preventDefault(), event.nativeEvent.stopImmediatePropagation(), event.target.click())
		: null;
