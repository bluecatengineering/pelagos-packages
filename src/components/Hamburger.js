import PropTypes from 'prop-types';

import handleButtonKeyDown from '../functions/handleButtonKeyDown';

import './Hamburger.less';

/** A hamburger menu. */
const Hamburger = ({active, onClick, ...props}) => (
	<div {...props} className="Hamburger" role="button" tabIndex="0" onClick={onClick} onKeyDown={handleButtonKeyDown}>
		<div className={'Hamburger__inner' + (active ? ' Hamburger--active' : '')} />
	</div>
);

Hamburger.propTypes = {
	/** Whether the button is active. */
	active: PropTypes.bool,
	/** Function invoked when the button is clicked. */
	onClick: PropTypes.func,
};

export default Hamburger;
