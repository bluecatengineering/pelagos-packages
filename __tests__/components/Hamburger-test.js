import React from 'react';
import {shallow} from 'enzyme';

import Hamburger from '../../src/components/Hamburger';
import handleButtonKeyDown from '../../src/functions/handleButtonKeyDown';

jest.unmock('../../src/components/Hamburger');

describe('Hamburger', () => {
	describe('rendering', () => {
		it('renders expected elements', () => {
			const wrapper = shallow(<Hamburger active />);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when active is not set', () => {
			const wrapper = shallow(<Hamburger />);
			expect(wrapper.getElement()).toMatchSnapshot();
		});
	});

	describe('behaviour', () => {
		it('calls onClick when the button is clicked', () => {
			const onClick = jest.fn();
			const wrapper = shallow(<Hamburger onClick={onClick} />);
			wrapper.find('[role="button"]').simulate('click');
			expect(onClick).toHaveBeenCalledTimes(1);
		});

		it('calls the key handler when a key is pressed', () => {
			const event = {keyCode: 13};
			const wrapper = shallow(<Hamburger />);
			wrapper.find('[role="button"]').simulate('keydown', event);
			expect(handleButtonKeyDown.mock.calls).toEqual([[event]]);
		});
	});
});
