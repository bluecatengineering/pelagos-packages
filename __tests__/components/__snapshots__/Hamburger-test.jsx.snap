// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hamburger rendering renders expected elements 1`] = `
<div
  className="Hamburger"
  role="button"
  tabIndex="0"
>
  <div
    className="Hamburger__inner Hamburger--active"
  />
</div>
`;

exports[`Hamburger rendering renders expected elements when active is not set 1`] = `
<div
  className="Hamburger"
  role="button"
  tabIndex="0"
>
  <div
    className="Hamburger__inner"
  />
</div>
`;
